src/algebra/jordan/triple.lean

```diff
@@ -0,0 +1,173 @@
+add_monoid_hom.map_zsmul ⟨λ (a : A), ⦃a, b, c⦄, lzero b c, λ _ _, ladd _ _ _ _⟩ _ _
+
+lemma mzsmul (a b c : A) (z : ℤ) : ⦃a, z•b, c⦄ = z•⦃a, b, c⦄ :=
+add_monoid_hom.map_zsmul ⟨λ (b : A), ⦃a, b, c⦄, mzero a c, λ _ _, madd _ _ _ _⟩ _ _
+
+lemma rzsmul (a b : A) (z : ℤ) (c : A) : ⦃a, b, z•c⦄ = z•⦃a, b, c⦄ :=
+  by rw [comm, lzsmul, comm]
+
+lemma lneg (a b c : A) : ⦃-a, b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←ladd, neg_add_self, lzero]
+
+lemma mneg (a b c : A): ⦃a, -b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←madd, neg_add_self, mzero]
+
+lemma rneg (a b c : A): ⦃a, b, -c⦄ = -⦃a, b, c⦄ := by rw [comm, lneg, comm]
+
+lemma lsub (a b c d : A) : ⦃a - d, b, c⦄ = ⦃a, b, c⦄ - ⦃d, b, c⦄ :=
+by rw [eq_sub_iff_add_eq, ← ladd, sub_add_cancel]
+
+lemma msub (a b c d : A) : ⦃a, b - c, d⦄ = ⦃a, b, d⦄ - ⦃a, c, d⦄ :=
+by rw [eq_sub_iff_add_eq, ← madd, sub_add_cancel]
+
+
+lemma rsub (a b c d : A) : ⦃a, b, c - d⦄ = ⦃a, b, c⦄ - ⦃a, b, d⦄ :=
+by rw [comm, lsub, comm, comm d]
+
+lemma lr_bilinear (a₁ a₂ b c₁ c₂ : A) : ⦃a₁ + a₂, b, c₁ + c₂⦄ =
+  ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ :=
+calc ⦃a₁ + a₂, b, c₁ + c₂⦄ = ⦃a₁, b, c₁ + c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw ladd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + (⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄) : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ : by rw ← add_assoc
+
+lemma polar (a b c : A) : ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ :=
+calc ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃c, b, a⦄ + ⦃c, b, c⦄ :
+  by rw lr_bilinear a c b a c
+... = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw comm c b a
+... = ⦃a, b, a⦄ + (⦃a, b, c⦄ + ⦃a, b, c⦄)  + ⦃c, b, c⦄ : by rw ← add_assoc
+... = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw two_nsmul
+
+/-- The triple product as an additive monoid homomorphism in each variable -/
+@[simps] def add_monoid_hom.tp : A →+ A →+ A →+ A :=
+{ to_fun := λ a,
+  { to_fun := λ b,
+    { to_fun := λ c, ⦃a, b, c⦄,
+      map_zero' := by rw rzero,
+      map_add' := λ _ _, by rw radd, },
+    map_zero' := add_monoid_hom.ext $ λ b, begin
+      simp only [add_monoid_hom.zero_apply, add_monoid_hom.coe_mk],
+      rw mzero,
+    end,
+    map_add' := λ a₁ a₂, add_monoid_hom.ext $ λ b, begin
+      simp only [add_monoid_hom.coe_mk, add_monoid_hom.add_apply],
+      rw madd,
+    end, },
+  map_zero' := add_monoid_hom.ext $ λ b, add_monoid_hom.ext $ λ c, begin
+    simp only [add_monoid_hom.zero_apply, add_monoid_hom.coe_mk],
+    rw lzero,
+  end,
+  map_add' := λ a₁ a₂, add_monoid_hom.ext $ λ b, add_monoid_hom.ext $ λ c, begin
+    simp only [add_monoid_hom.coe_mk, add_monoid_hom.add_apply],
+    rw ladd,
+  end, }
+
+/-- Define the multiplication operator `D` -/
+@[simps] def D : A →+ A →+ add_monoid.End A := add_monoid_hom.tp
+
+/-- Define the quadratic operator `Q` -/
+@[simps] def Q : A →+ A →+  add_monoid.End A :=
```

Comment:
Is this `D.comp add_monoid_hom.flip`?

Reply:
We have:
```
lemma Q_def (a b c : A) : Q a c b = ⦃a, b, c⦄ := by simp
```
But 
```
def Q' : A →+ A →+  add_monoid.End A := D.comp add_monoid_hom.flip
```
errors with:
```
type mismatch at application
  D.comp add_monoid_hom.flip
term
  add_monoid_hom.flip
has type
  (?m_1 →+ ?m_2 →+ ?m_3) → ?m_2 →+ ?m_1 →+ ?m_3 : Type (max (max ? ? ?) ? ? ?)
but is expected to have type
  A →+ A : Type u_1
```

Reply:
How about `D.comp (add_monoid_hom.flip : (A →+ A →+ A) →(A →+ A →+ A) `?

Reply:
Ah nevermind, I see the problem. Does `add_monoid_hom.flip_hom` exist?

Reply:
This appears to work `def Q : A →+ A →+  add_monoid.End A := (D : A →+ A →+ add_monoid.End A).flip_hom`.

Reply:
Spoke too soon. That gives me `Q a c b = ⦃c, a, b⦄` rather than `Q a c b = ⦃a, b, c⦄`.

Reply:
I would expect you want `D.comp add_monoid_hom.flip_hom`, with type annotations if lean doesn't like that.